### PR TITLE
Switch gradient checkpointing default to use_reentrant=False (PyTorch recommended)

### PR DESCRIPTION
**Pull request by qgallouedec**
### Set `use_reentrant=False` by default 

This PR defaults gradient checkpointing to `use_reentrant=False` in TRL when no explicit value is provided.

PyTorch now recommends the non-reentrant checkpointing variant, see https://docs.pytorch.org/docs/stable/checkpoint.html. However, Transformers still defaults to `use_reentrant=True` because it was explicitly pinned in the past to silence a PyTorch warning during a transition period, and the default was never updated afterward.

Until this is fixed upstream and released (see https://github.com/huggingface/transformers/pull/43203), TRL aligns with the current PyTorch recommendation by setting `use_reentrant=False` by default, while fully preserving any user-provided value.

### Fixes`Expected to mark a variable ready only once`

This PR fixes an issue that seems unrelated, but is: https://github.com/huggingface/trl/issues/4782

### Remove `ScriptArguments.gradient_checkpointing_use_reentrant`

`ScriptArguments.gradient_checkpointing_use_reentrant` exists but is never used. This is misleading, so this PR removes this argument.

### CODE REVIEW

File: trl/trainer/dpo_config.py [hunk 1]
```diff
@@ -500,6 +500,14 @@ class DPOConfig(TrainingArguments):
...
     def __post_init__(self):
         self.bf16 = not (self.fp16) if self.bf16 is None else self.bf16
+
+        # Transformers explicitly set use_reentrant=True in the past to silence a PyTorch warning, but the default was
+        # never updated once PyTorch switched to recommending use_reentrant=False. Until that change lands upstream
+        # (see https://github.com/huggingface/transformers/pull/43203) and is released (most likely in 5.0.0), we
+        # default to the recommended non-reentrant behavior here, while preserving any user-provided value.
+        self.gradient_checkpointing_kwargs = self.gradient_checkpointing_kwargs or {}
+        self.gradient_checkpointing_kwargs.setdefault("use_reentrant", False)
+
```
Comment (albertvillanova) [inline trl/trainer/dpo_config.py:513]: First, what about setting an explicit condition on the transformers versions where this setting is necessary? That would help maintainability by making clear when this PR is no longer necessary and can be reverted in the future.

Second, does it make sense to have `self.gradient_checkpointing_kwargs = {"use_reentrant": False}` even if the user passed `gradient_checkpointing=False`?

Comment (qgallouedec) [reply trl/trainer/dpo_config.py:513]: Good point, addressed in 269189c35b1c6c47c2c623afe09c6a402c85cb65

**Comment by qgallouedec** [comment]
> Additionally, there are 2 new CI errors that I think we should check if they are caused by this PR and then if they should be fixed as well:

Fixed, `enable_input_require_grads()` is needed not only for reentrant.
